apps/astarte_pairing/lib/astarte_pairing/fdo/service_info.ex

```
@@ -0,0 +1,87 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2025 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.Pairing.FDO.ServiceInfo do
+  alias COSE.Messages.Encrypt0
+
+  @owner_max_service_info 4096
+
+  def handle_msg_66(raw_cose_body, sevk, old_voucher, session_cipher) do
+    with {:ok, %{payload: payload, uhdr: uhdr, phdr: phdr}} <-
+           Encrypt0.decrypt_decode(raw_cose_body, session_cipher, sevk),
+         {:ok, [replacement_hmac, _device_max_size], _} <- CBOR.decode(payload),
+         {:ok, _new_voucher} <-
+           generate_replacement_voucher(old_voucher, replacement_hmac) do
+      # TODO: Store `new_voucher` and `device_max_size` in the Session or DB.
+
+      msg_67_payload = [@owner_max_service_info]
+      current_iv = Map.get(uhdr, :iv)
```

this allows current_iv to be nil, which would violently fail on generate_msg_67

instead, you could add to the with block something like

```elixir
{:ok, %CBOR.Tag{tag: :bytes, value: iv}} <- Map.fetch(uhdr, :iv)
```

to fail fast in this function. remember: all validation should be performed asap and you should have your core functionality safe from the outside world